### PR TITLE
Detect network issues from state commands

### PR DESCRIPTION
**Pull request by MDrakos**
https://www.pivotaltracker.com/story/show/174609354

### CODE REVIEW

**Comment by MDrakos** [comment]
This is an example of the output screen after these changes:

image

File: installers/msi-language/ActiveState/Network.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        /// <summary>
+        /// ResetErrorDetails clears the registry entries for network errors.
+        /// This function must be run from a deferred custom action
+        /// </summary>
+        public static void ResetErrorDetails(Session session)
+        {
+            // Deferred custom actions run as administrator so we have to specifically set
+            // the registry key for the user using their SID in order for the value to
+            // be available in later immediate custom actions
+            string registryKey = string.Format("HKEY_USERS\\{0}\\SOFTWARE\\ActiveState\\{1}", session.CustomActionData["USERSID"], session.CustomActionData["PRODUCT_NAME"]);
```
Comment (Naatan) [inline installers/msi-language/ActiveState/Network.cs:22]: Is `USERSID` still valid given @mdrohmann's latest changes?

File: internal/retryhttp/client.go [hunk 2]
```diff
@@ -71,19 +66,19 @@ func normalizeResponse(res *http.Response, err error) (*http.Response, error) {
 	if res != nil {
 		switch res.StatusCode {
 		case 408:
-			return res, locale.WrapInputError(&UserNetworkError{408}, "err_user_network_server_timeout", "Request failed due to timeout during communication with server. {{.V0}}", solutionLocale)
+			return res, locale.WrapInputError(&UserNetworkError{408}, "err_user_network_server_timeout", "Request failed due to timeout during communication with server.")
 		case 425:
-			return res, locale.WrapInputError(&UserNetworkError{425}, "err_user_network_tooearly", "Request failed due to retrying connection too fast. {{.V0}}", solutionLocale)
+			return res, locale.WrapInputError(&UserNetworkError{425}, "err_user_network_tooearly", "Request failed due to retrying connection too fast.")
 		case 429:
-			return res, locale.WrapInputError(&UserNetworkError{429}, "err_user_network_toomany", "Request failed due to too many requests. {{.V0}}", solutionLocale)
```
Comment (Naatan) [inline internal/retryhttp/client.go:78]: These are here for the CLI use-case. Not the end of the world if we repeat this in the MSI.

**Review by Naatan — CHANGES_REQUESTED** [review]

File: installers/msi-language/CustomFatalError.wxs [hunk 3]
```diff
@@ -18,13 +18,13 @@
 				<Control Id="CustomTextFatalError" Type="Hyperlink" X="135" Y="110" Width="220" Height="20" Transparent="yes" Hidden="yes" Text="!(loc.FatalErrorPrimary)">
 					<Condition Action="show">NETWORK_ERROR = "false"</Condition>
 				</Control>
-				<Control Id="CustomTextNetworkError" Type="Text" X="135" Y="70" Width="220" Height="60" Transparent="yes" Hidden="yes" Text="!(loc.NetworkErrorPrimary)">
+				<Control Id="CustomTextNetworkError" Type="Text" X="135" Y="60" Width="220" Height="60" Transparent="yes" Hidden="yes" Text="!(loc.NetworkErrorPrimary)">
 					<Condition Action="show">NETWORK_ERROR = "true"</Condition>
 				</Control>
-				<Control Id="CustomTextNetworkErrorDetails" Type="Text" X="135" Y="130" Width="220" Height="40" Transparent="yes" Hidden="yes" Text="!(loc.NetworkErrorDetails)">
+				<Control Id="CustomTextNetworkErrorDetails" Type="Text" X="135" Y="120" Width="220" Height="60" Transparent="yes" Hidden="yes" Text="!(loc.NetworkErrorDetails)">
```
Comment (mdrohmann) [inline installers/msi-language/CustomFatalError.wxs:24]: I wonder if we can actually use a [ScrollableText control](https://docs.microsoft.com/en-us/windows/win32/msi/scrollabletext-control) here, just like in the EULA dialog.  That way we would ensure that we always have enough space for the error text.

File: installers/msi-language/ActiveState/Network.cs (hunk 1 above)
Comment (mdrohmann) [reply installers/msi-language/ActiveState/Network.cs:22]: I believe that `USERSID` would still be the right choice here.

**Review by mdrohmann — CHANGES_REQUESTED** [review]
As we cannot control the length of the error message very well, I thought it would be nice to use a ScrollableText element for it.

File: installers/msi-language/ActiveState/Network.cs (hunk 1 above)
Comment (MDrakos) [reply installers/msi-language/ActiveState/Network.cs:22]: Yes, the `USERSID` stuff is still required here so we can pass the network error details from the deferred custom action to an immediate one

File: installers/msi-language/CustomFatalError.wxs (hunk 3 above)
Comment (MDrakos) [reply installers/msi-language/CustomFatalError.wxs:24]: I spent some time trying different alternatives. The issue with a `ScrollableText` element is that it requires an `.rft` file at build time to populate the text box. Apparently it is possible to open the MSI database at runtime and replace that file, however after some investigation I did not go down that route. Come to think of it I don't think a deferred custom action can access the MSI database as it has to get access via the `session` object.

File: internal/retryhttp/client.go (hunk 2 above)
Comment (Naatan) [reply internal/retryhttp/client.go]: That's fine, this'll give them the information they need.

**Review by Naatan — DISMISSED** [review]

File: installers/msi-language/CustomFatalError.wxs (hunk 3 above)
Comment (mdrohmann) [reply installers/msi-language/CustomFatalError.wxs:24]: I actually also remember looking into that for some other dialog.  Well, another of those strange limitations.  Let's leave it like this.  There is quite a bit of space after all...

**Review by mdrohmann — DISMISSED** [review]

**Review by mdrohmann — APPROVED** [review]

**Review by Naatan — APPROVED** [review]

**Pull request merged**